cluster_management/src/main/java/com/pinterest/rocksplicator/eventstore/LeaderEventHistoryStore.java

```diff
@@ -0,0 +1,498 @@
+import com.pinterest.rocksplicator.thrift.eventhistory.LeaderEventType;
+import com.pinterest.rocksplicator.thrift.eventhistory.LeaderEventsHistory;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import com.google.common.cache.RemovalListener;
+import com.google.common.cache.RemovalNotification;
+import com.google.common.math.IntMath;
+import org.apache.curator.framework.CuratorFramework;
+import org.apache.curator.framework.CuratorFrameworkFactory;
+import org.apache.curator.retry.ExponentialBackoffRetry;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.BiConsumer;
+
+public class LeaderEventHistoryStore implements Closeable {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(LeaderEventHistoryStore.class);
+  private static final int NUM_PARALLEL_THREADS = 16;
+
+  private final String zkConnectString;
+  private final String clusterName;
+  private final CuratorFramework zkClient;
+  private final Optional<Integer> maxEventsToKeep;
+  private final Codec<LeaderEventsHistory, byte[]> leaderEventsHistoryCodec;
+  private final List<ExecutorService> executorServices;
+
+  private final LoadingCache<String, LoadingCache<String,
+      MergeableReadWriteStore<LeaderEventsHistory, LeaderEvent>>>
+      zkStoreCache;
+
+  private final LoadingCache<String, Cache<String, LeaderEventsHistory>> localHistoryCache;
+
+  private final List<Lock> clusterResourcePartitionLock;
+
+  public LeaderEventHistoryStore(
+      final String zkConnectString,
+      final String clusterName,
+      final Optional<Integer> maxEventsToKeep) {
+    this.zkConnectString = Preconditions.checkNotNull(zkConnectString);
+    this.clusterName = Preconditions.checkNotNull(clusterName);
+
+    this.leaderEventsHistoryCodec = new WrappedDataThriftCodec(
+        LeaderEventsHistory.class, SerializationProtocol.COMPACT, CompressionAlgorithm.GZIP);
+    this.maxEventsToKeep = maxEventsToKeep;
+
+    this.zkClient =
+        CuratorFrameworkFactory.newClient(Preconditions.checkNotNull(this.zkConnectString),
+            new ExponentialBackoffRetry(1000, 3));
+
+    this.zkClient.start();
+    try {
+      this.zkClient.blockUntilConnected(60, TimeUnit.SECONDS);
+    } catch (InterruptedException e) {
+      LOGGER.error(String.format("Can't connect to zookeeper: %s", zkConnectString), e);
```

Comment:
Should we be continuing even if connection to zk is not successful?

Reply:
No, if the zk provided is something we can't connect to when the process start... we can terminate the process instead... making sure correct zk is provided for storing leader events.. else it shouldn't be provided at all. Throwing runtime exception in this case.